### ISSUE TITLE
a11y scrolling broken with traversal order

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#14987]
Looking at the following app:

```dart
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new CustomScrollView(
        slivers: [
          new SliverAppBar(
            title: new Text('Hello'),
          ),
          new SliverGrid.count(
              crossAxisCount: 2,
              children: new List.generate(100, (int i) {
                return new Semantics(
                  container: true,
                  child: new Container(
                      padding: const EdgeInsets.all(16.0),
                      decoration: new BoxDecoration(
                        border: new Border.all(color: Colors.green),
                      ),
                      child: new Text("Item $i")
                  ),
                );
              })
          ),
        ],
      ),
    );
  }
}
```

It produces a Grid layout (2 items in a row). Scroll down a bit and printing the semantics tree gives us this:

```
I/flutter ( 3311): SemanticsNode#0(Rect.fromLTRB(0.0, 0.0, 1440.0, 2392.0))
I/flutter ( 3311): └SemanticsNode#22(Rect.fromLTRB(0.0, 0.0, 411.4, 683.4) scaled by 3.5x, nextNodeId: 30, previousNodeId: 36)
I/flutter ( 3311):  └SemanticsNode#30(Rect.fromLTRB(0.0, 0.0, 411.4, 683.4), actions: [scrollDown, scrollUp], nextNodeId: 37, previousNodeId: 22, scrollExtentMin: 0.0, scrollPosition: 1148.1, scrollExtentMax: 9682.3)
I/flutter ( 3311):   ├SemanticsNode#35(Rect.fromLTRB(0.0, -39.5, 205.7, 166.2), label: "Item 10", textDirection: ltr, nextNodeId: 36, previousNodeId: -1)
I/flutter ( 3311):   ├SemanticsNode#36(Rect.fromLTRB(205.7, -39.5, 411.4, 166.2), label: "Item 11", textDirection: ltr, nextNodeId: 22, previousNodeId: 35)
I/flutter ( 3311):   ├SemanticsNode#37(Rect.fromLTRB(0.0, 166.2, 205.7, 371.9), label: "Item 12", textDirection: ltr, nextNodeId: 38, previousNodeId: 30)
I/flutter ( 3311):   ├SemanticsNode#38(Rect.fromLTRB(205.7, 166.2, 411.4, 371.9), label: "Item 13", textDirection: ltr, nextNodeId: 39, previousNodeId: 37)
I/flutter ( 3311):   ├SemanticsNode#39(Rect.fromLTRB(0.0, 371.9, 205.7, 577.6), label: "Item 14", textDirection: ltr, nextNodeId: 40, previousNodeId: 38)
I/flutter ( 3311):   ├SemanticsNode#40(Rect.fromLTRB(205.7, 371.9, 411.4, 577.6), label: "Item 15", textDirection: ltr, nextNodeId: 41, previousNodeId: 39)
I/flutter ( 3311):   ├SemanticsNode#41(Rect.fromLTRB(0.0, 577.6, 205.7, 783.3), label: "Item 16", textDirection: ltr, nextNodeId: 42, previousNodeId: 40)
I/flutter ( 3311):   └SemanticsNode#42(Rect.fromLTRB(205.7, 577.6, 411.4, 783.3), label: "Item 17", textDirection: ltr, nextNodeId: -1, previousNodeId: 41)
I/flutter ( 3311): 
```

It is expected that you can traverse the items in the grid one-by-one, giving the following order: 35 <> 36 <> 37 <> 38 <> 39 <> 40 <> 41 <> 42
As seen in the tree the actual traversal order is: 35 <> 36 <> **22 <> 30** <> 37 <> 38 <> 39 <> 40 <> 42 <> 42

The nodes 22 and 30 do not belong in there and they break Android a11y scrolling behavior. For example: If a11y focus is on node 38 and you swipe left (because you want to go to node 37) **Android issues an incorrect scrolling command causing a jump in the scroll position** (and then ends up with focus on node 37 because the nodes in between are not focusable). The incorrect scrolling command is probably issued because Android tries to follow the specified traversal order which would lead you outside of the scrollable container. Whenever this happens and the container is still scrollable, Android issues an implicit scroll command.

/cc @gspencergoog

**Comment by goderbauer** [comment]
Possibly related: https://github.com/flutter/flutter/issues/14681

(I believe the traversal order is calculated incorrectly because the first few nodes are half outside of the enclosing scrollable container)

**Comment by goderbauer** [comment]
There is also an issue when you move the a11y focus from the last visible node (e.g. 42 in the example above) to the next currently hidden node: This should scroll up one page and then focus on the next node. Instead, we currently scroll up two pages.
